### ISSUE TITLE
Bug: C# Flipt client implementation fails to match on "Entity" constraints

### ISSUE THREAD

**Issue by ian-a-anderson** [flipt-io/flipt-client-sdks#595]
Problem
---

We have been using [the "Entity" type](https://github.com/flipt-io/flipt/pull/2779) when defining various constraints for our segments, and this has been working as expected when accessing Flipt over GRPC. We recently switched to this [newer client SDK](https://github.com/flipt-io/flipt-client-sdks) for all the various advantages that it brings, but noticed that there appears to be a bug that prevents matching when using this specific constraint type. 

Repro
---

### 1. Create a segment and define a single constraint using the "Entity" type, EG:

<img width="560" alt="image" src="https://github.com/user-attachments/assets/c04f25ff-f278-433d-bbac-3da7de9ed344" />

<img width="1351" alt="image" src="https://github.com/user-attachments/assets/5e91a8a9-9675-480d-9277-f01a050af631" />

### 2. Create a new flag and apply this segment as a rule:

<img width="1351" alt="image" src="https://github.com/user-attachments/assets/5e91a8a9-9675-480d-9277-f01a050af631" />

<img width="1326" alt="image" src="https://github.com/user-attachments/assets/c6b2b011-4fc8-49c4-a822-9bec1f1df0b5" />

### 3. Note that evaluating this rule in the console succeeds:
 
<img width="1296" alt="image" src="https://github.com/user-attachments/assets/ce452f62-79ae-4386-aa4d-69dc2d71cfa2" />

### 4. ISSUE - when passing the same information up via the new Flipt client SDK (in our case, [the C# SDK](https://github.com/flipt-io/flipt-client-sdks/tree/main/flipt-client-csharp)) this flag fails to match. 

<img width="1775" alt="image" src="https://github.com/user-attachments/assets/0d461abb-83d0-444e-ad3b-483a5a513825" />

**Comment by erka** [comment]
Thank you @ian-a-anderson for a bug report.  Which versions of Flipt and client do you use?

**Comment by markphelps** [comment]
@ian-a-anderson do you see anything in the Flipt Server logs like:

```
skipping `entity_id` constraint type to support older client; upgrade client to allow `entity_id` constraint type
```
?

**Comment by ian-a-anderson** [comment]
@erka This is Flipt v1.45.0 and C# client v0.1.0

**Comment by ian-a-anderson** [comment]
@markphelps no I don't see anything like that in the logs. All I see from today are periodic "cleanup process deleting authentications" entries.

**Comment by erka** [comment]
@ian-a-anderson I think you face this bug flipt-io/flipt#3261 which was fixed in Flipt v1.45.1. Could you please upgrade Flipt to that version and try to repo your issue?

**Comment by ian-a-anderson** [comment]
Thanks @erka, we are in the process of upgrading our server instance and I'll ping back here if it's working as expected on the latest backend.

**Comment by ian-a-anderson** [comment]
> [@ian-a-anderson](https://github.com/ian-a-anderson) do you see anything in the Flipt Server logs like:
> 
> ```
> skipping `entity_id` constraint type to support older client; upgrade client to allow `entity_id` constraint type
> ```
> 
> ?

@markphelps some better log tooling has exposed that we are seeing these errors. Still having gotten through the server upgrade yet but I assume this will resolve once the server is up-to-date.

**Comment by markphelps** [comment]
@ian-a-anderson yes this should resolve after updating the server to a more recent version (v1.45.1 or above)

**Comment by ian-a-anderson** [comment]
Updating the server to v1.54.2 has resolved this for us, thanks!

**Comment by tjjevans1** [comment]
Seeing a similar issue with the browser SDK.

Log:

"2025-02-06T14:31:19Z\t\u001b[33mWARN\u001b[0m\tskipping `entity_id` constraint type to support older client; upgrade client to allow `entity_id` constraint type\t{\"server\": \"grpc\", \"namespace\": \"audio-staging\", \"flag\": \"transcriptions\"}"

"@flipt-io/flipt-client-browser": "^0.4.0".

Flipt version v1.54.2.

**Comment by erka** [comment]
> Seeing a similar issue with the browser SDK.

@tjjevans1 I can't repo it. It works fine for me locally. Do you have any ideas or steps how to face your problem?

**Comment by tjjevans1** [comment]
I've got quite a complex use case.

I'm using a self hosted instance of Flipt as part of a Kubernetes cluster deployed via helm chart.

I've got a proxy sitting in front of that which passes requests onto the Flipt server. That's already using the Authentication header, so I have to include my flipt token as a custom header, and reassign in the proxy.

In order to pass the custom header with the request, I've defined a custom fetcher, but I didn't realise that the default fetcher is passing some custom headers already, namely `x-flipt-accept-server-version` which I suspect is needed for the `entity_id` feature described above.

Will try adding that header to my fetcher and see if that works.

**Comment by tjjevans1** [comment]
Confirmed, adding the header to my custom fetcher allowed me to use the entity constraint.

Are there any examples available for using custom fetchers?

**Comment by erka** [comment]
I believe the only one example is the default fetcher. https://github.com/flipt-io/flipt-client-sdks/blob/50065c1014d92b917b20e49e0f55d454ef8ae4cd/flipt-client-browser/src/index.ts#L78-L93

**Comment by tjjevans1** [comment]
Ah thanks, yes this is what I used for reference during development.

I got confused because the function depends on the `url` and `header` variables defined higher in the scope.

Are there plans to make this a pure function?

**Comment by erka** [comment]
@tjjevans1 You're the first who raises this. We are open for suggestions. Probably it will be a breaking change but I don't think that many people have custom fetcher implementation.
